sql/tests: TestRandomSyntaxSQLSmith failed [inner-join RelExpr children have intersecting columns]

sql/tests.TestRandomSyntaxSQLSmith [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/12747385?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/12747385?buildTab=artifacts#/) on release-23.2 @ [c9f6f30496fe34af8d17410e5307f359458aaf52](https://github.com/cockroachdb/cockroach/commits/c9f6f30496fe34af8d17410e5307f359458aaf52):

Random syntax error:

```
    rsg_test.go:911: Crash detected: server panic: pq: internal error: inner-join RelExpr children have intersecting columns: (42,50,63,64)
```
Query:

```
        SELECT
        	'1976-11-23':::DATE AS " col885550",
        	var_pop(tab391036.crdb_internal_mvcc_timestamp::DECIMAL)::DECIMAL AS col885551,
        	stddev((-0.07257373532581902993):::DECIMAL::INT8)::DECIMAL AS col͇885552,
        	tab391036."co l1_0" AS col885553
        FROM
        	defaultdb.public.seed@[0] AS "%qtab391035"
        	JOIN defaultdb.public."tab'le1"@"tab'le1_expr_idx" AS tab391036 ON
        			("%qtab391035".tableoid) = (tab391036.tableoid)
        			AND ("%qtab391035".crdb_internal_mvcc_timestamp) = (tab391036.crdb_internal_mvcc_timestamp)
        WHERE
        	(SELECT 884958838:::OID AS " col885549" FROM defaultdb.public."tab'le1"@"tab'le1_expr_co l1_0_idx" AS "_�tab391040", defaultdb.public.seed@[0] AS "ta""b391041" ORDER BY "ta""b391041"._int2 DESC NULLS FIRST, "ta""b391041"._interval NULLS FIRST LIMIT 1:::INT8)::OID
        	IN (
        			WITH
        				with159481 (col885545)
        					AS (
        						SELECT * FROM (VALUES (e'\U00074537' COLLATE de_DE), (NULL)) AS tab391037 (col885545)
        						INTERSECT ALL
        							SELECT
        								*
        							FROM
        								(
        									VALUES
        										(text((-876764555.4284667560):::DECIMAL::DECIMAL)::STRING COLLATE de_DE),
        										(e'\U0008831A' COLLATE de_DE),
        										(e'\U0005AFE8\U0008DCBA\U000D391A' COLLATE de_DE)
        								)
        									AS tab391038 (col885546)
        					),
        				with159482 ("coL885547")
        					AS (
        						SELECT * FROM (VALUES ('1988-09-26':::DATE), (NULL), (NULL)) AS "tab\\uDC40391039" ("coL885547")
        					)
        			SELECT
        				tab391036.tableoid::OID AS c̃ol̼😜885548
        		)
        GROUP BY
        	"%qtab391035"._decimal, "%qtab391035"._date, tab391036.crdb_internal_mvcc_timestamp, tab391036."co l1_0"
        HAVING
        	bool_and("%qtab391035"._bool::BOOL)::BOOL
        ORDER BY
        	"%qtab391035"._decimal ASC, "%qtab391035"._date DESC NULLS FIRST
        LIMIT
        	26:::INT8;
```
Schema:

```
    rsg_test.go:720: To reproduce, use schema:
    rsg_test.go:722: SET sql_safe_updates = false;;
    rsg_test.go:722: SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false;;
    rsg_test.go:722: SET CLUSTER SETTING sql.stats.histogram_collection.enabled = false;;
    rsg_test.go:722: CREATE TABLE "tab'le1" ("co l1_0" NAME, col1_1 TIMETZ NOT NULL, "c ol1_2" INT2 NOT NULL, col1_3 STRING NOT NULL AS (lower("co l1_0")) VIRTUAL, PRIMARY KEY (col1_3 ASC, "c ol1_2" ASC), INDEX (("c ol1_2" + 17774:::INT8) ASC, "co l1_0" DESC), UNIQUE (col1_1 DESC, lower(CAST(col1_1 AS STRING)) ASC, col1_3) WHERE "tab'le1"."co l1_0" < '':::STRING:::NAME, INDEX (col1_1 ASC, ("c ol1_2" + (-26953):::INT8), col1_3 ASC) WHERE (("tab'le1".col1_3 != 'X':::STRING) AND ("tab'le1"."c ol1_2" = 1:::INT8)) OR ("tab'le1".col1_1 < '00:00:00+15:59':::TIMETZ) VISIBILITY 0.30, INDEX (lower("co l1_0") DESC) STORING (col1_1) WHERE ((("tab'le1".col1_1 != '24:00:00-15:59':::TIMETZ) OR ("tab'le1"."co l1_0" <= 'X':::STRING:::NAME)) OR ("tab'le1".col1_3 < e'\U00002603':::STRING)) OR ("tab'le1"."c ol1_2" >= (-32768):::INT8), UNIQUE ("c ol1_2" DESC, "co l1_0" ASC) WHERE ("tab'le1".col1_1 > '24:00:00-15:59':::TIMETZ) AND ("tab'le1"."co l1_0" = e'\x00':::STRING:::NAME), UNIQUE ("c ol1_2" ASC, col1_3 ASC, col1_1 ASC), FAMILY ("c ol1_2", col1_1, "co l1_0"));
    rsg_test.go:722: ALTER TABLE "tab'le1" INJECT STATISTICS e'[{"avg_size": 23, "columns": ["\\"co l1_0\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 881177514894877516, "histo_col_type": "", "name": "__auto__", "null_count": 1382871461899841653, "row_count": 2482386356606809074}, {"avg_size": 25, "columns": ["col1_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1325107802613261019, "histo_buckets": [{"distinct_range": 0, "num_eq": 0, "num_range": 0, "upper_bound": "09:32:44.209818+06:54"}, {"distinct_range": 455092794166317200, "num_eq": 300, "num_range": 4025135702079162861, "upper_bound": "05:28:20.158276+00:32"}, {"distinct_range": 0, "num_eq": 3015553724450640402, "num_range": 7561892057817281519, "upper_bound": "18:44:36.635038+12:41"}, {"distinct_range": 2303590125519348500, "num_eq": 5664224463233182992, "num_range": 2303590125519348549, "upper_bound": "06:49:58.228979+00:11"}, {"distinct_range": 3000000000, "num_eq": 3000000000, "num_range": 3000000000, "upper_bound": "22:15:53.024518-07:15"}, {"distinct_range": 0, "num_eq": 523283071033843517, "num_range": 9019577961766724534, "upper_bound": "24:00:00-15:59"}], "histo_col_type": "TIMETZ", "name": "__auto__", "null_count": 0, "row_count": 2482386356606809074}, {"avg_size": 12, "columns": ["\\"c ol1_2\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 703602587645114629, "histo_col_type": "INT2", "name": "__auto__", "null_count": 0, "row_count": 2482386356606809074}, {"avg_size": 7, "columns": ["col1_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 325547840088071858, "histo_buckets": [{"distinct_range": 0, "num_eq": 4794843170231017593, "num_range": 0, "upper_bound": ""}, {"distinct_range": 458.1811842957447, "num_eq": 1333789801729865774, "num_range": 600, "upper_bound": "\\u0000"}, {"distinct_range": 25146.153969309224, "num_eq": 2792342344167939272, "num_range": 40000, "upper_bound": "XN"}], "histo_col_type": "STRING", "name": "__auto__", "null_count": 0, "row_count": 2482386356606809074}]':::JSONB;
    rsg_test.go:722: CREATE TYPE rand_typ_0 AS ENUM ('bjsaf', 'xxbcdn', 'mobyq', 'b', 'zvzwzh');
    rsg_test.go:722: CREATE TYPE rand_typ_1 AS ENUM ('w', 'ahy', 'lnc');
    rsg_test.go:722: SET sql_safe_updates = false;;
    rsg_test.go:722: 
        BEGIN; CREATE TYPE greeting AS ENUM ('hello', 'howdy', 'hi', 'good day', 'morning'); COMMIT;
        BEGIN;
        CREATE TABLE IF NOT EXISTS seed AS
        	SELECT
        		g::INT2 AS _int2,
        		g::INT4 AS _int4,
        		g::INT8 AS _int8,
        		g::FLOAT4 AS _float4,
        		g::FLOAT8 AS _float8,
        		'2001-01-01'::DATE + g AS _date,
        		'2001-01-01'::TIMESTAMP + g * '1 day'::INTERVAL AS _timestamp,
        		'2001-01-01'::TIMESTAMPTZ + g * '1 day'::INTERVAL AS _timestamptz,
        		g * '1 day'::INTERVAL AS _interval,
        		g % 2 = 1 AS _bool,
        		g::DECIMAL AS _decimal,
        		g::STRING AS _string,
        		g::STRING::BYTES AS _bytes,
        		substring('00000000-0000-0000-0000-' || g::STRING || '00000000000', 1, 36)::UUID AS _uuid,
        		'0.0.0.0'::INET + g AS _inet,
        		g::STRING::JSONB AS _jsonb,
        		enum_range('hello'::greeting)[g] as _enum
        	FROM
        		generate_series(1, 5) AS g;
        COMMIT;
        
        INSERT INTO seed DEFAULT VALUES;
        CREATE INDEX on seed (_int8, _float8, _date);
        CREATE INVERTED INDEX on seed (_jsonb);
        ;
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (8594355245580624359);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (7814313793381395637);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (452728068839437155);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (8289981896376721186);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (1102571324373106807);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (9120260909951863919);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (1368567861377401943);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (505761125082940084);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (3764695915680452516);
    rsg_test.go:722: ALTER TABLE seed SCATTER;
    rsg_test.go:724: 
    rsg_test.go:725: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d437d2c847dfedbc4972f231c3331c8e/logTestRandomSyntaxSQLSmith513922758
--- FAIL: TestRandomSyntaxSQLSmith (377.67s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSQLSmith.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-33641

--- a/pkg/sql/opt/metadata.go
+++ b/pkg/sql/opt/metadata.go
@@ -115,6 +115,10 @@ type Metadata struct {
 	// mutation operators, used to determine the logical properties of WithScan.
 	withBindings map[WithID]Expr
 
+	// hoistedUncorrelatedSubqueries is used to track uncorrelated subqueries
+	// that have been hoisted.
+	hoistedUncorrelatedSubqueries map[Expr]struct{}
+
 	// dataSourceDeps stores each data source object that the query depends on.
 	dataSourceDeps map[cat.StableID]cat.DataSource
 
@@ -1060,6 +1064,24 @@ func (md *Metadata) ForEachWithBinding(fn func(WithID, Expr)) {
 	for id, expr := range md.withBindings {
 		fn(id, expr)
 	}
+}
+
+// AddHoistedUncorrelatedSubquery marks the given uncorrelated subquery
+// expression as hoisted. It is used to prevent hoisting the same uncorrelated
+// subquery twice because that may cause two children of an expression to have
+// intersecting columns (see #114703).
+func (md *Metadata) AddHoistedUncorrelatedSubquery(subquery Expr) {
+	if md.hoistedUncorrelatedSubqueries == nil {
+		md.hoistedUncorrelatedSubqueries = make(map[Expr]struct{})
+	}
+	md.hoistedUncorrelatedSubqueries[subquery] = struct{}{}
+}
+
+// IsHoistedUncorrelatedSubquery returns true if the given subquery was
+// previously marked as hoisted with AddHoistedUncorrelatedSubquery.
+func (md *Metadata) IsHoistedUncorrelatedSubquery(subquery Expr) bool {
+	_, ok := md.hoistedUncorrelatedSubqueries[subquery]
+	return ok
 }
 
 // TestingDataSourceDeps exposes the dataSourceDeps for testing.

--- a/pkg/sql/opt/norm/decorrelate_funcs.go
+++ b/pkg/sql/opt/norm/decorrelate_funcs.go
@@ -68,15 +68,22 @@ func (c *CustomFuncs) deriveHasHoistableSubquery(scalar opt.ScalarExpr) bool {
 		return false
 
 	case *memo.EqExpr:
-		// Hoist subqueries in expressions like (Eq (Variable) (Subquery)) if
-		// the corresponding session setting is enabled.
+		// Hoist subqueries in expressions like (Eq (Variable) (Subquery)) if:
+		//
+		//   1. The corresponding session setting is enabled.
+		//   2. And, the subquery has not already been hoisted elsewhere in the
+		//      expression tree. Hoisting the same subquery twice could result
+		//      in query plans where two children of an expression have
+		//      intersecting columns (see #114703).
+		//
 		// TODO(mgartner): We could hoist if we have an IS NOT DISTINCT FROM
 		// expression. But it won't currently lead to a lookup join due to
 		// #100855 and the plan could be worse, so we avoid it for now.
 		if c.f.evalCtx.SessionData().OptimizerHoistUncorrelatedEqualitySubqueries {
 			_, isLeftVar := scalar.Child(0).(*memo.VariableExpr)
-			_, isRightSubquery := scalar.Child(1).(*memo.SubqueryExpr)
-			if isLeftVar && isRightSubquery {
+			subquery, isRightSubquery := scalar.Child(1).(*memo.SubqueryExpr)
+			if isLeftVar && isRightSubquery &&
+				!c.f.Metadata().IsHoistedUncorrelatedSubquery(subquery) {
 				return true
 			}
 		}
@@ -824,12 +831,18 @@ func (r *subqueryHoister) hoistAll(scalar opt.ScalarExpr) opt.ScalarExpr {
 		// According to the implementation of deriveHasHoistableSubquery,
 		// Exists, Any, and ArrayFlatten expressions are only hoistable if they
 		// are correlated. Uncorrelated subquery expressions are hoistable if
-		// the corresponding session setting is enabled and they are part of an
-		// equality expression with a variable.
-		uncorrelatedHoistAllowed := scalar.Op() == opt.SubqueryOp &&
-			r.f.evalCtx.SessionData().OptimizerHoistUncorrelatedEqualitySubqueries
-		if subquery.Relational().OuterCols.Empty() && !uncorrelatedHoistAllowed {
-			break
+		// the corresponding session setting is enabled, they are part of an
+		// equality expression with a variable, and they have not already been
+		// hoisted elsewhere in the expression tree.
+		if subquery.Relational().OuterCols.Empty() {
+			uncorrelatedHoistAllowed := scalar.Op() == opt.SubqueryOp &&
+				r.f.evalCtx.SessionData().OptimizerHoistUncorrelatedEqualitySubqueries &&
+				!r.f.Metadata().IsHoistedUncorrelatedSubquery(scalar)
+			if !uncorrelatedHoistAllowed {
+				break
+			}
+			// Mark the subquery as being hoisted.
+			r.f.Metadata().AddHoistedUncorrelatedSubquery(scalar)
 		}
 
 		switch t := scalar.(type) {

--- a/pkg/sql/opt/norm/testdata/rules/decorrelate
+++ b/pkg/sql/opt/norm/testdata/rules/decorrelate
@@ -4847,6 +4847,238 @@ select
                           └── max [as=max:12, outer=(8)]
                                └── x:8
 
+# The subquery should only be hoisted once to avoid creating an expression with
+# children that have intersecting columns. See #114703.
+norm expect=HoistSelectSubquery
+SELECT NULL
+FROM a AS t1
+JOIN a AS t2 ON t1.i = t2.i
+WHERE t2.i = (SELECT 0 FROM a)
+----
+project
+ ├── columns: "?column?":23
+ ├── fd: ()-->(23)
+ ├── inner-join (hash)
+ │    ├── columns: t1.i:2!null t2.i:9!null "?column?":22!null
+ │    ├── fd: ()-->(2,9,22), (2)==(9,22), (22)==(2,9), (9)==(2,22)
+ │    ├── inner-join (hash)
+ │    │    ├── columns: t1.i:2!null "?column?":22!null
+ │    │    ├── multiplicity: left-rows(zero-or-one), right-rows(zero-or-more)
+ │    │    ├── fd: ()-->(2,22), (2)==(22), (22)==(2)
+ │    │    ├── scan a [as=t1]
+ │    │    │    └── columns: t1.i:2
+ │    │    ├── max1-row
+ │    │    │    ├── columns: "?column?":22!null
+ │    │    │    ├── error: "more than one row returned by a subquery used as an expression"
+ │    │    │    ├── cardinality: [0 - 1]
+ │    │    │    ├── key: ()
+ │    │    │    ├── fd: ()-->(22)
+ │    │    │    └── project
+ │    │    │         ├── columns: "?column?":22!null
+ │    │    │         ├── fd: ()-->(22)
+ │    │    │         ├── scan a
+ │    │    │         └── projections
+ │    │    │              └── 0 [as="?column?":22]
+ │    │    └── filters
+ │    │         └── t1.i:2 = "?column?":22 [outer=(2,22), constraints=(/2: (/NULL - ]; /22: (/NULL - ]), fd=(2)==(22), (22)==(2)]
+ │    ├── select
+ │    │    ├── columns: t2.i:9!null
+ │    │    ├── scan a [as=t2]
+ │    │    │    └── columns: t2.i:9
+ │    │    └── filters
+ │    │         └── eq [outer=(9), subquery, constraints=(/9: (/NULL - ])]
+ │    │              ├── t2.i:9
+ │    │              └── subquery
+ │    │                   └── max1-row
+ │    │                        ├── columns: "?column?":22!null
+ │    │                        ├── error: "more than one row returned by a subquery used as an expression"
+ │    │                        ├── cardinality: [0 - 1]
+ │    │                        ├── key: ()
+ │    │                        ├── fd: ()-->(22)
+ │    │                        └── project
+ │    │                             ├── columns: "?column?":22!null
+ │    │                             ├── fd: ()-->(22)
+ │    │                             ├── scan a
+ │    │                             └── projections
+ │    │                                  └── 0 [as="?column?":22]
+ │    └── filters
+ │         └── t1.i:2 = t2.i:9 [outer=(2,9), constraints=(/2: (/NULL - ]; /9: (/NULL - ]), fd=(2)==(9), (9)==(2)]
+ └── projections
+      └── NULL [as="?column?":23]
+
+# Each subquery should only be hoisted once.
+norm expect=HoistSelectSubquery
+SELECT NULL
+FROM a AS t1
+JOIN a AS t2 ON t1.i = t2.i
+WHERE t2.i = (SELECT 0 FROM a) AND t1.i = (SELECT 0 FROM a)
+----
+project
+ ├── columns: "?column?":31
+ ├── fd: ()-->(31)
+ ├── inner-join (hash)
+ │    ├── columns: t1.i:2!null t2.i:9!null "?column?":22!null "?column?":30!null
+ │    ├── fd: ()-->(2,9,22,30), (2)==(9,22,30), (30)==(2,9,22), (22)==(2,9,30), (9)==(2,22,30)
+ │    ├── inner-join (hash)
+ │    │    ├── columns: t1.i:2!null "?column?":22!null "?column?":30!null
+ │    │    ├── multiplicity: left-rows(zero-or-one), right-rows(zero-or-more)
+ │    │    ├── fd: ()-->(2,22,30), (2)==(22,30), (30)==(2,22), (22)==(2,30)
+ │    │    ├── inner-join (hash)
+ │    │    │    ├── columns: t1.i:2!null "?column?":30!null
+ │    │    │    ├── multiplicity: left-rows(zero-or-one), right-rows(zero-or-more)
+ │    │    │    ├── fd: ()-->(2,30), (2)==(30), (30)==(2)
+ │    │    │    ├── scan a [as=t1]
+ │    │    │    │    └── columns: t1.i:2
+ │    │    │    ├── max1-row
+ │    │    │    │    ├── columns: "?column?":30!null
+ │    │    │    │    ├── error: "more than one row returned by a subquery used as an expression"
+ │    │    │    │    ├── cardinality: [0 - 1]
+ │    │    │    │    ├── key: ()
+ │    │    │    │    ├── fd: ()-->(30)
+ │    │    │    │    └── project
+ │    │    │    │         ├── columns: "?column?":30!null
+ │    │    │    │         ├── fd: ()-->(30)
+ │    │    │    │         ├── scan a
+ │    │    │    │         └── projections
+ │    │    │    │              └── 0 [as="?column?":30]
+ │    │    │    └── filters
+ │    │    │         └── t1.i:2 = "?column?":30 [outer=(2,30), constraints=(/2: (/NULL - ]; /30: (/NULL - ]), fd=(2)==(30), (30)==(2)]
+ │    │    ├── select
+ │    │    │    ├── columns: "?column?":22!null
+ │    │    │    ├── cardinality: [0 - 1]
+ │    │    │    ├── key: ()
+ │    │    │    ├── fd: ()-->(22)
+ │    │    │    ├── max1-row
+ │    │    │    │    ├── columns: "?column?":22!null
+ │    │    │    │    ├── error: "more than one row returned by a subquery used as an expression"
+ │    │    │    │    ├── cardinality: [0 - 1]
+ │    │    │    │    ├── key: ()
+ │    │    │    │    ├── fd: ()-->(22)
+ │    │    │    │    └── project
+ │    │    │    │         ├── columns: "?column?":22!null
+ │    │    │    │         ├── fd: ()-->(22)
+ │    │    │    │         ├── scan a
+ │    │    │    │         └── projections
+ │    │    │    │              └── 0 [as="?column?":22]
+ │    │    │    └── filters
+ │    │    │         └── eq [outer=(22), subquery, constraints=(/22: (/NULL - ])]
+ │    │    │              ├── "?column?":22
+ │    │    │              └── subquery
+ │    │    │                   └── max1-row
+ │    │    │                        ├── columns: "?column?":30!null
+ │    │    │                        ├── error: "more than one row returned by a subquery used as an expression"
+ │    │    │                        ├── cardinality: [0 - 1]
+ │    │    │                        ├── key: ()
+ │    │    │                        ├── fd: ()-->(30)
+ │    │    │                        └── project
+ │    │    │                             ├── columns: "?column?":30!null
+ │    │    │                             ├── fd: ()-->(30)
+ │    │    │                             ├── scan a
+ │    │    │                             └── projections
+ │    │    │                                  └── 0 [as="?column?":30]
+ │    │    └── filters
+ │    │         └── t1.i:2 = "?column?":22 [outer=(2,22), constraints=(/2: (/NULL - ]; /22: (/NULL - ]), fd=(2)==(22), (22)==(2)]
+ │    ├── select
+ │    │    ├── columns: t2.i:9!null
+ │    │    ├── scan a [as=t2]
+ │    │    │    └── columns: t2.i:9
+ │    │    └── filters
+ │    │         ├── eq [outer=(9), subquery, constraints=(/9: (/NULL - ])]
+ │    │         │    ├── t2.i:9
+ │    │         │    └── subquery
+ │    │         │         └── max1-row
+ │    │         │              ├── columns: "?column?":22!null
+ │    │         │              ├── error: "more than one row returned by a subquery used as an expression"
+ │    │         │              ├── cardinality: [0 - 1]
+ │    │         │              ├── key: ()
+ │    │         │              ├── fd: ()-->(22)
+ │    │         │              └── project
+ │    │         │                   ├── columns: "?column?":22!null
+ │    │         │                   ├── fd: ()-->(22)
+ │    │         │                   ├── scan a
+ │    │         │                   └── projections
+ │    │         │                        └── 0 [as="?column?":22]
+ │    │         └── eq [outer=(9), subquery, constraints=(/9: (/NULL - ])]
+ │    │              ├── t2.i:9
+ │    │              └── subquery
+ │    │                   └── max1-row
+ │    │                        ├── columns: "?column?":30!null
+ │    │                        ├── error: "more than one row returned by a subquery used as an expression"
+ │    │                        ├── cardinality: [0 - 1]
+ │    │                        ├── key: ()
+ │    │                        ├── fd: ()-->(30)
+ │    │                        └── project
+ │    │                             ├── columns: "?column?":30!null
+ │    │                             ├── fd: ()-->(30)
+ │    │                             ├── scan a
+ │    │                             └── projections
+ │    │                                  └── 0 [as="?column?":30]
+ │    └── filters
+ │         └── t1.i:2 = t2.i:9 [outer=(2,9), constraints=(/2: (/NULL - ]; /9: (/NULL - ]), fd=(2)==(9), (9)==(2)]
+ └── projections
+      └── NULL [as="?column?":31]
+
+exec-ddl
+CREATE FUNCTION f114703() RETURNS INT STABLE LANGUAGE SQL AS $$
+  SELECT x FROM xy
+$$
+----
+
+# The filter "t2.a = f()" is pushed into both sides of the join, and inlined as
+# a subquery that is only hoisted once.
+norm expect=HoistSelectSubquery
+SELECT NULL
+FROM a AS t1
+JOIN a AS t2 ON t1.i = t2.i
+WHERE t2.i = f114703()
+----
+project
+ ├── columns: "?column?":19
+ ├── fd: ()-->(19)
+ ├── inner-join (hash)
+ │    ├── columns: t1.i:2!null t2.i:9!null x:15!null
+ │    ├── fd: ()-->(2,9,15), (2)==(9,15), (15)==(2,9), (9)==(2,15)
+ │    ├── inner-join (hash)
+ │    │    ├── columns: t1.i:2!null x:15!null
+ │    │    ├── multiplicity: left-rows(zero-or-one), right-rows(zero-or-more)
+ │    │    ├── fd: ()-->(2,15), (2)==(15), (15)==(2)
+ │    │    ├── scan a [as=t1]
+ │    │    │    └── columns: t1.i:2
+ │    │    ├── limit
+ │    │    │    ├── columns: x:15!null
+ │    │    │    ├── cardinality: [0 - 1]
+ │    │    │    ├── key: ()
+ │    │    │    ├── fd: ()-->(15)
+ │    │    │    ├── scan xy
+ │    │    │    │    ├── columns: x:15!null
+ │    │    │    │    ├── key: (15)
+ │    │    │    │    └── limit hint: 1.00
+ │    │    │    └── 1
+ │    │    └── filters
+ │    │         └── t1.i:2 = x:15 [outer=(2,15), constraints=(/2: (/NULL - ]; /15: (/NULL - ]), fd=(2)==(15), (15)==(2)]
+ │    ├── select
+ │    │    ├── columns: t2.i:9!null
+ │    │    ├── scan a [as=t2]
+ │    │    │    └── columns: t2.i:9
+ │    │    └── filters
+ │    │         └── eq [outer=(9), subquery, constraints=(/9: (/NULL - ])]
+ │    │              ├── t2.i:9
+ │    │              └── subquery
+ │    │                   └── limit
+ │    │                        ├── columns: x:15!null
+ │    │                        ├── cardinality: [0 - 1]
+ │    │                        ├── key: ()
+ │    │                        ├── fd: ()-->(15)
+ │    │                        ├── scan xy
+ │    │                        │    ├── columns: x:15!null
+ │    │                        │    ├── key: (15)
+ │    │                        │    └── limit hint: 1.00
+ │    │                        └── 1
+ │    └── filters
+ │         └── t1.i:2 = t2.i:9 [outer=(2,9), constraints=(/2: (/NULL - ]; /9: (/NULL - ]), fd=(2)==(9), (9)==(2)]
+ └── projections
+      └── NULL [as="?column?":19]
+
 
 # --------------------------------------------------
 # HoistProjectSubquery

